src/main/java/org/opensearch/searchrelevance/ml/AdaptiveRateLimiter.java

```diff
@@ -0,0 +1,238 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.searchrelevance.ml;
+
+import java.util.Locale;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.searchrelevance.ml.connector.ConnectorType;
+
+/**
+ * Adaptive rate limiter that learns optimal rates per model and handles circuit breaking
+ */
+public class AdaptiveRateLimiter {
```

Comment:
can you please add abstraction layer and extract interface for rate limiter, similar to what you have done for LLMConnector. I suggest humble interface name `RateLimiter`